src/batch/src/executor2/project.rs

```diff
@@ -56,7 +56,11 @@ impl ProjectExecutor2 {
                 .iter_mut()
                 .map(|expr| expr.eval(&data_chunk).map(Column::new))
                 .collect::<Result<Vec<_>>>()?;
-            let ret = DataChunk::builder().columns(arrays).build();
+            let ret = if arrays.is_empty() {
+                DataChunk::new_dummy(data_chunk.cardinality())
+            } else {
+                DataChunk::builder().columns(arrays).build()
```

Comment:
If we have `.visibility()` when building this new data chunk, then we don't need the change in `HashShuffleSender`. Do I understand this correctly?

Reply:
Actually, there are two bug fix in this pr, although there related to one issue.
The code you reference is to make Project deal with zero columns non-zero cardinality.

Reply:
I see, can we still let the executor instead of `ShuffleSender` decide whether to compact a data chunk? `BatchHashAgg` would produce incorrect results?

Reply:
Actually it is the Exchange that produce incorrect results.